File: ios/RNLPInboxMessage.m
//
//  RNLPInboxMessage.m
//  RNLeanplum
//
//  Created by Sky Eckstrom on 9/28/18.
//  Copyright © 2018 Branding Brand. All rights reserved.
//

#import "RNLPInboxMessage.h"
#import "LPInboxMessage+RNAdditions.h"

#if __has_include(<Leanplum/Leanplum.h>)
#import <Leanplum/Leanplum.h>
#elif __has_include("Leanplum.h")
#import "Leanplum.h"
#else
#import "Leanplum/Leanplum.h"
#endif

static NSString * const kRNLPInboxMessageNotFoundErrorCode = @"RNLPInboxMessageNotFound";
static NSString * const kRNLPInboxMessageNotFoundErrorReason = @"Could not find a message with the given id.";

@implementation RNLPInboxMessage
RCT_EXPORT_MODULE(LPInboxMessage);

RCT_EXPORT_METHOD(readMessageId:(NSString *)messageId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    LPInboxMessage *message = [[Leanplum inbox] messageForId:messageId];
    
    if (message) {
        [message read];

        return resolve([NSNull null]);
    }

    return reject(kRNLPInboxMessageNotFoundErrorCode, kRNLPInboxMessageNotFoundErrorReason, nil);

}

RCT_EXPORT_METHOD(removeMessageId:(NSString *)messageId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    LPInboxMessage *message = [[Leanplum inbox] messageForId:messageId];
    
    if (message) {
        [message remove];

        return resolve([NSNull null]);
    }

    return reject(kRNLPInboxMessageNotFoundErrorCode, kRNLPInboxMessageNotFoundErrorReason, nil);
}

@end

File: android/src/main/java/com/reactnativeleanplum/RNLPInboxMessage.java
package com.reactnativeleanplum;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import android.app.Application;

import com.leanplum.Leanplum;
import com.leanplum.LeanplumInbox;
import com.leanplum.LeanplumInboxMessage;

public class RNLPInboxMessage extends ReactContextBaseJavaModule {
    private static final String E_MESSAGE_NOT_FOUND_ERROR = "RNLPInboxMessageNotFound";
    private static final String E_MESSAGE_NOT_FOUND_REASON = "Could not find a message with the given id.";

    public RNLPInboxMessage(ReactApplicationContext reactContext, Application app) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "LPInboxMessage";
    }

    @ReactMethod
    public void readMessageId(String messageId, Promise promise) {
        LeanplumInboxMessage message = Leanplum.getInbox().messageForId(messageId);

        if (message != null) {
            message.read();

            promise.resolve(null);
        } else {
            promise.reject(E_MESSAGE_NOT_FOUND_ERROR, E_MESSAGE_NOT_FOUND_REASON);
        }
    }

    @ReactMethod
    public void removeMessageId(String messageId, Promise promise) {
        LeanplumInboxMessage message = Leanplum.getInbox().messageForId(messageId);

        if (message != null) {
            message.remove();

            promise.resolve(null);
        } else {
            promise.reject(E_MESSAGE_NOT_FOUND_ERROR, E_MESSAGE_NOT_FOUND_REASON);
        }
    }
}
